<h1>Section 1:  Use DLG Inc Ecomm data set for this section<h1>

In [2]:
import pandas as pd
import collections

In [3]:
cd C:\Users\danny\AB Data Scientist TakeHome


C:\Users\danny\AB Data Scientist TakeHome


In [4]:
df = pd.read_csv (r'C:\Users\danny\AB Data Scientist TakeHome\DLG_INC_EComm_Data.csv')

In [5]:
df.head() 

event_time event_type  product_id          category_id  \
0  2019-10-01 00:00:05 UTC       view     1480613  2053013561092866816   
1  2019-10-01 00:00:10 UTC       view    28719074  2053013565480108800   
2  2019-10-01 00:00:13 UTC       view     3900746  2053013552326770944   
3  2019-10-01 00:00:23 UTC       view     1004739  2053013555631882240   
4  2019-10-01 00:00:31 UTC       view     3900746  2053013552326770944   

                         category_code   brand   price    user_id  \
0                    computers.desktop  pulser  908.62  512742880   
1                   apparel.shoes.keds   baden  102.71  520571932   
2  appliances.environment.water_heater   haier  102.38  555444559   
3               electronics.smartphone  xiaomi  197.55  519530528   
4  appliances.environment.water_heater   haier  102.38  555444559   

                           user_session  
0  0d0d91c2-c9c2-4e81-90a5-86594dec0db9  
1  ac1cd4e5-a3ce-4224-a2d7-ff660a105880  
2  98b88fa0-d8fa-4b9d-8a71-3dd403afab85  
3  9882d21f-2c5f-496b-90d4-a1503edb6562  
4  98b88fa0-d8fa-4b9d-8a71-3dd403afab85

In [6]:
df.columns

Index(['event_time', 'event_type', 'product_id', 'category_id',
       'category_code', 'brand', 'price', 'user_id', 'user_session'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2656063 entries, 0 to 2656062
Data columns (total 9 columns):
 #   Column         Dtype  
---  ------         -----  
 0   event_time     object 
 1   event_type     object 
 2   product_id     int64  
 3   category_id    int64  
 4   category_code  object 
 5   brand          object 
 6   price          float64
 7   user_id        int64  
 8   user_session   object 
dtypes: float64(1), int64(3), object(5)
memory usage: 182.4+ MB


In [8]:
# Check the NAN value
df['event_type'].isnull().sum().sum()

0

In [9]:
# check what kind of categorical data in this column
set(df['event_type'].values)

{'cart', 'purchase', 'view'}

In [10]:
# conversion_rate
freq = df['event_type'].value_counts()
freq


view        2519917
cart          81190
purchase      54956
Name: event_type, dtype: int64

In [11]:
# conversion_rate
count_purchase =  freq[2]
count_view = freq[0]
conversion_rate = (count_purchase / count_view)
print('site conversion rate :' + str(conversion_rate*100) + '%')


site conversion rate :2.1808654808868706%


<h2>1.	What is site conversion rate (defined as purchase/view)<h2>

<h4>Based on the above result, site conversion rate :2.1808654808868706%<h4>

In [12]:
# Average User value
# step1 - Average Order Value

#len(set(df['user_id'].values))

In [13]:
df

event_time event_type  product_id          category_id  \
0        2019-10-01 00:00:05 UTC       view     1480613  2053013561092866816   
1        2019-10-01 00:00:10 UTC       view    28719074  2053013565480108800   
2        2019-10-01 00:00:13 UTC       view     3900746  2053013552326770944   
3        2019-10-01 00:00:23 UTC       view     1004739  2053013555631882240   
4        2019-10-01 00:00:31 UTC       view     3900746  2053013552326770944   
...                          ...        ...         ...                  ...   
2656058  2019-10-31 23:59:03 UTC       view    43300122  2097815753423060736   
2656059  2019-10-31 23:59:23 UTC       view     1003461  2053013555631882240   
2656060  2019-10-31 23:59:33 UTC       view     1005238  2053013555631882240   
2656061  2019-10-31 23:59:44 UTC       view     1004497  2053013555631882240   
2656062  2019-10-31 23:59:56 UTC       view     1003936  2053013555631882240   

                               category_code   brand   price    user_id  \
0                          computers.desktop  pulser  908.62  512742880   
1                         apparel.shoes.keds   baden  102.71  520571932   
2        appliances.environment.water_heater   haier  102.38  555444559   
3                     electronics.smartphone  xiaomi  197.55  519530528   
4        appliances.environment.water_heater   haier  102.38  555444559   
...                                      ...     ...     ...        ...   
2656058                       apparel.tshirt    nike   27.03  545220871   
2656059               electronics.smartphone  xiaomi  489.07  520088904   
2656060               electronics.smartphone    oppo  308.45  514171705   
2656061               electronics.smartphone   nokia  160.08  566143627   
2656062               electronics.smartphone  xiaomi  437.33  520088904   

                                 user_session  
0        0d0d91c2-c9c2-4e81-90a5-86594dec0db9  
1        ac1cd4e5-a3ce-4224-a2d7-ff660a105880  
2        98b88fa0-d8fa-4b9d-8a71-3dd403afab85  
3        9882d21f-2c5f-496b-90d4-a1503edb6562  
4        98b88fa0-d8fa-4b9d-8a71-3dd403afab85  
...                                       ...  
2656058  f278cca0-e0f6-49a3-819a-d961998282d5  
2656059  4d3b30da-a5e4-49df-b1a8-ba5943f1dd33  
2656060  168955ad-db0f-4f0f-b71b-5dd8cfc9e113  
2656061  aa610ab3-5c60-4551-8a2b-8c999dddf843  
2656062  4d3b30da-a5e4-49df-b1a8-ba5943f1dd33  

[2656063 rows x 9 columns]

In [14]:
condition = (df['event_type'] == 'purchase')

grouped = df[condition].reset_index()
grouped

index               event_time event_type  product_id  \
0           31  2019-10-01 00:07:07 UTC   purchase    13800054   
1          140  2019-10-01 02:21:02 UTC   purchase     1004659   
2          183  2019-10-01 02:23:25 UTC   purchase     1005135   
3          196  2019-10-01 02:23:59 UTC   purchase     1004833   
4          281  2019-10-01 02:26:57 UTC   purchase     1004566   
...        ...                      ...        ...         ...   
54951  2655768  2019-10-31 23:18:52 UTC   purchase     1004839   
54952  2655872  2019-10-31 23:33:11 UTC   purchase     4804056   
54953  2655900  2019-10-31 23:36:41 UTC   purchase     1004863   
54954  2655914  2019-10-31 23:38:20 UTC   purchase     5100562   
54955  2655981  2019-10-31 23:48:20 UTC   purchase     1005244   

               category_id                category_code    brand    price  \
0      2053013557418656000    furniture.bathroom.toilet  santeri    54.42   
1      2053013555631882240       electronics.smartphone  samsung   787.18   
2      2053013555631882240       electronics.smartphone    apple  1747.79   
3      2053013555631882240       electronics.smartphone  samsung   174.76   
4      2053013555631882240       electronics.smartphone   huawei   177.20   
...                    ...                          ...      ...      ...   
54951  2053013555631882240       electronics.smartphone     oppo   179.12   
54952  2053013554658803712  electronics.audio.headphone    apple   160.57   
54953  2053013555631882240       electronics.smartphone  samsung   170.14   
54954  2053013553341792256           electronics.clocks    apple   298.60   
54955  2053013555631882240       electronics.smartphone     vivo   463.31   

         user_id                          user_session  
0      555332717  1dea3ee2-2ded-42e8-8e7a-4e2ad6ae942f  
1      555168263  d068d08f-aaf3-4949-b031-cdaf49a5c7cf  
2      515384420  7f82b450-6c45-4346-96fb-ecf4ab25779c  
3      524102669  0cf768dd-f1d3-4108-b06f-75868aacf6b0  
4      518962372  718c8069-2782-4585-acad-6747f6f2f79d  
...          ...                                   ...  
54951  566092490  763b3e3f-a5c7-4f90-8f39-ada3220304ff  
54952  526208213  d175757c-15ef-40a8-a0aa-03d3ee90e065  
54953  559242302  371c4d85-d114-4da4-9f17-d902ea47f7ca  
54954  519366802  52329a85-fe91-4635-842b-66d3e819b92a  
54955  519641980  3c70d8b4-6c15-4938-8e90-66f76ca85857  

[54956 rows x 10 columns]

In [15]:
#grouped.groupby(['user_id','price']).sum().head(120)
#df2 = grouped.groupby(['user_id','price'])
#len(df2)

In [16]:
grouped.sort_values(by=['user_id'], ascending=True)

index               event_time event_type  product_id  \
32600  1542016  2019-10-18 19:15:13 UTC   purchase     1003304   
50327  2412890  2019-10-28 16:36:26 UTC   purchase     9100708   
32315  1521928  2019-10-18 15:43:17 UTC   purchase     5100572   
25990  1261325  2019-10-15 20:23:14 UTC   purchase     1004781   
52848  2537662  2019-10-30 11:04:15 UTC   purchase     1005252   
...        ...                      ...        ...         ...   
54896  2652648  2019-10-31 19:50:40 UTC   purchase     1004870   
54901  2652871  2019-10-31 19:56:53 UTC   purchase     1004767   
54908  2653195  2019-10-31 20:14:11 UTC   purchase     4804295   
54922  2654153  2019-10-31 20:55:15 UTC   purchase     1004870   
54923  2654222  2019-10-31 20:58:19 UTC   purchase     1004739   

               category_id                category_code    brand   price  \
32600  2053013555631882240       electronics.smartphone    apple  577.94   
50327  2053013552888807936  computers.peripherals.mouse    apple  127.42   
32315  2053013553341792256           electronics.clocks    apple  437.85   
25990  2053013555631882240       electronics.smartphone   huawei  271.75   
52848  2053013555631882240       electronics.smartphone     oppo  308.53   
...                    ...                          ...      ...     ...   
54896  2053013555631882240       electronics.smartphone  samsung  275.25   
54901  2053013555631882240       electronics.smartphone  samsung  242.63   
54908  2053013554658803712  electronics.audio.headphone   xiaomi   22.80   
54922  2053013555631882240       electronics.smartphone  samsung  275.25   
54923  2053013555631882240       electronics.smartphone   xiaomi  203.09   

         user_id                          user_session  
32600  403013066  10456526-1e4c-487b-824b-04dd0d1b73d1  
50327  418115333  c3dc5500-9390-4ff9-86f6-8e7a2801a9b2  
32315  435786120  d40d4861-902f-4e70-8801-b3a28183ec15  
25990  446080337  010cf56f-1769-48a3-b904-12eb681532e8  
52848  447698613  27431636-2368-4b03-81a5-bbfcea2bd471  
...          ...                                   ...  
54896  566240064  1d1d9d9f-9381-4be6-b396-75c71c284f96  
54901  566241984  c35a4567-006a-462a-b640-43731403cc50  
54908  566243643  963f7384-c88e-4722-9932-777f232dd95a  
54922  566254545  8d27dd17-b8fd-4623-ad32-77228d969424  
54923  566254639  8c41d7b9-48ad-4800-a16a-4701a21b2df8  

[54956 rows x 10 columns]

In [17]:
Total_revenue = grouped['price'].sum()
Number_of_Orders = len(grouped.groupby(['user_id','price']))
AOV = Total_revenue / Number_of_Orders # or Average Purchase Value

In [18]:
AOV

381.6868921970866

In [19]:
#Purchase Frequency
uniq_customers = len(set(grouped['user_id']))
PF = Number_of_Orders/ uniq_customers #Average Purchase Freqeuncy Rate


In [20]:
uniq_customers

45226

In [21]:
PF 

1.1748551718038296

In [22]:
#User_value
user_val= AOV * PF
print('User value is ' + str(user_val))

User value is 448.42681930747796


<h2>2.What is the average user value?<h2>
<h4>Average Conversion value = 448.42681930747796<h4>

In [23]:
# Conversion value
# My assumption: Profit margin: 35 % / Closing ratio 20%
# Sales revenue * Closing ratio * profit margin
avg_revenue = Total_revenue/Number_of_Orders
closing_ratio = 0.2
profit_margin = 0.35

In [24]:
avg_conversion = avg_revenue * closing_ratio * profit_margin
print('Average Conversion value = ' + str(avg_conversion))

Average Conversion value = 26.71808245379606


In [25]:
# Conversion value 2
# Revenue / # of conversion
# # of conversion = number of payments/transaction

avg_revenue2 = Total_revenue
n_of_transaction = len(grouped)
avg_conversion2 = Total_revenue /n_of_transaction
print('Average Conversion value = ' + str(avg_conversion2))

Average Conversion value = 369.0325229274328


<h2>3.	What is average conversion value?<h2>
<h4>Average Conversion value = 369.0325229274328<h4>

In [26]:
# check unique category
#category_list = set(df['category_code'])
category_list = df['category_code'].unique()


In [27]:
d = collections.defaultdict(list)
df3 = df[['event_type', 'category_code']]
df3

event_type                        category_code
0             view                    computers.desktop
1             view                   apparel.shoes.keds
2             view  appliances.environment.water_heater
3             view               electronics.smartphone
4             view  appliances.environment.water_heater
...            ...                                  ...
2656058       view                       apparel.tshirt
2656059       view               electronics.smartphone
2656060       view               electronics.smartphone
2656061       view               electronics.smartphone
2656062       view               electronics.smartphone

[2656063 rows x 2 columns]

In [28]:
d

defaultdict(list, {})

In [29]:
#condition = (df['event_type'] == 'view')
for i in range(len(df3)):
    cat = df.loc[i, 'category_code']
    d[cat] = [0,0,0]

In [30]:
for i in range(len(df3)):
    event = df3.loc[i, 'event_type']
    cat = df.loc[i, 'category_code']
    if event == 'view':
        d[cat][0] += 1
    if event == 'purchase':
        d[cat][1] += 1


In [31]:
for cat, val in d.items():
    d[cat][2] = (val[1] / val[0]) * 100

    

In [32]:
d

defaultdict(list,
            {'computers.desktop': [39439, 297, 0.7530616902051269],
             'apparel.shoes.keds': [35708, 267, 0.7477316007617341],
             'appliances.environment.water_heater': [13233,
              275,
              2.0781379883624274],
             'electronics.smartphone': [1060091, 33801, 3.18849985520111],
             'computers.peripherals.monitor': [10465, 116, 1.1084567606306737],
             'computers.notebook': [110126, 1617, 1.4683181083486188],
             'appliances.environment.air_heater': [12525,
              200,
              1.5968063872255487],
             'appliances.environment.vacuum': [74580,
              1213,
              1.6264414052024672],
             'kids.toys': [8065, 72, 0.8927464352138872],
             'apparel.shoes': [60211, 331, 0.5497334374117686],
             'appliances.kitchen.refrigerators': [68804,
              866,
              1.2586477530376141],
             'construction.tools.drill': [20947, 20

<h2>4.	Which are top 5 and bottom 5 categories in terms of conversion rate?<h2>

In [33]:
dic = sorted(d.items(), key = lambda x: x[1][2], reverse= True)
dic[:5]

[('kids.fmcg.diapers', [2189, 74, 3.380539058931019]),
 ('electronics.smartphone', [1060091, 33801, 3.18849985520111]),
 ('electronics.audio.headphone', [100622, 3035, 3.0162389934606746]),
 ('apparel.shorts', [35, 1, 2.857142857142857]),
 ('stationery.cartrige', [582, 14, 2.405498281786942])]

<h3># TOP 5 categories (For visualization, create new dataframe)<h3>

In [34]:

keys = [k for k,v in dic]
values = [v[2] for k,v in dic]
df3 = pd.DataFrame.from_dict({'index': keys, 'values': values})
df3.columns = ['category_code', 'Conversion_rate']
df3.head().style.hide_index()

<h3># Bottom 5 categories (For visualization, create new dataframe)<h3>

In [35]:
# Bottom 5 categories (For visualization, create new dataframe)
df3.tail().style.hide_index()


In [36]:
import matplotlib.pyplot as plt
import seaborn as sns

In [37]:
# add new column for the conversion rate
new_df = pd.merge(left=df, right=df3, on='category_code')

In [38]:
new_df

event_time event_type  product_id          category_id  \
0        2019-10-01 00:00:05 UTC       view     1480613  2053013561092866816   
1        2019-10-01 00:11:43 UTC       view     1480509  2053013561092866816   
2        2019-10-01 00:13:07 UTC       view     1401577  2053013561092866816   
3        2019-10-01 02:19:04 UTC       view     1480519  2053013561092866816   
4        2019-10-01 02:21:15 UTC       view     1480456  2053013561092866816   
...                          ...        ...         ...                  ...   
2656058  2019-10-31 09:50:33 UTC       view    55500000  2147123662858224384   
2656059  2019-10-31 09:57:30 UTC       view    55500002  2147123662858224384   
2656060  2019-10-31 13:06:42 UTC       view    55500001  2147123662858224384   
2656061  2019-10-31 15:03:32 UTC       view    55500001  2147123662858224384   
2656062  2019-10-31 17:08:36 UTC       view    55500001  2147123662858224384   

                        category_code   brand    price    user_id  \
0                   computers.desktop  pulser   908.62  512742880   
1                   computers.desktop  pulser   692.40  512742880   
2                   computers.desktop   apple  1415.48  542378517   
3                   computers.desktop  pulser   658.94  512742880   
4                   computers.desktop  pulser  1016.73  512742880   
...                               ...     ...      ...        ...   
2656058  auto.accessories.anti_freeze   wurth    25.74  516817782   
2656059  auto.accessories.anti_freeze   wurth     5.92  513353872   
2656060  auto.accessories.anti_freeze   wurth    23.04  516881079   
2656061  auto.accessories.anti_freeze   wurth    23.04  512995007   
2656062  auto.accessories.anti_freeze   wurth    23.04  543974567   

                                 user_session  Conversion_rate  
0        0d0d91c2-c9c2-4e81-90a5-86594dec0db9         0.753062  
1        1b7451ac-14e6-470f-a218-cdeb3e68dc21         0.753062  
2        673efa91-486f-4a00-8d76-510a6b7034df         0.753062  
3        dc4b4278-a623-4113-b76a-391e2b54e3ad         0.753062  
4        dc4b4278-a623-4113-b76a-391e2b54e3ad         0.753062  
...                                       ...              ...  
2656058  65dd513f-59ca-4766-9648-08974e019129         0.000000  
2656059  beeabe3b-b104-4ac4-9812-21ed7c059a1a         0.000000  
2656060  cf614338-bcd9-4d10-b4fe-c6844d6a6121         0.000000  
2656061  805928ca-a714-48c2-b92f-83a92ddf18a2         0.000000  
2656062  2c0e228f-18a3-4ae8-9a0d-f79374687889         0.000000  

[2656063 rows x 10 columns]

<h2>5.	What is top pair of categories which have high affinity for each other.<h2>

<h5>In here, I defined the the category affinity as 'A visitor’s activity in one session, such as which category they view most often, can be used for targeting in subsequent visits'.Each category page a visitor views during a session is captured, and his or her “favorite” category is calculated based on a recency and frequency model.
    Thus, I will caculate the cart/ view for each categories and extract the top pair of categories.<h5>

In [39]:
d2 = collections.defaultdict(list)
df4 = df[['event_type', 'category_code']]
df4

event_type                        category_code
0             view                    computers.desktop
1             view                   apparel.shoes.keds
2             view  appliances.environment.water_heater
3             view               electronics.smartphone
4             view  appliances.environment.water_heater
...            ...                                  ...
2656058       view                       apparel.tshirt
2656059       view               electronics.smartphone
2656060       view               electronics.smartphone
2656061       view               electronics.smartphone
2656062       view               electronics.smartphone

[2656063 rows x 2 columns]

In [46]:
for i in range(len(df4)):
    cat = df.loc[i, 'category_code']
    d2[cat] = [0,0,0] #[key : category_code, value :[view count, cart count, average  (cart/ view)]]

In [41]:
for i in range(len(df4)):
    event = df4.loc[i, 'event_type']
    cat = df4.loc[i, 'category_code']
    if event == 'view':
        d2[cat][0] += 1
    if event == 'cart':
        d2[cat][1] += 1


In [42]:
for cat, val in d2.items():
    d2[cat][2] = (val[1] / val[0]) * 100

    

In [43]:
d

defaultdict(list,
            {'computers.desktop': [39439, 297, 0.7530616902051269],
             'apparel.shoes.keds': [35708, 267, 0.7477316007617341],
             'appliances.environment.water_heater': [13233,
              275,
              2.0781379883624274],
             'electronics.smartphone': [1060091, 33801, 3.18849985520111],
             'computers.peripherals.monitor': [10465, 116, 1.1084567606306737],
             'computers.notebook': [110126, 1617, 1.4683181083486188],
             'appliances.environment.air_heater': [12525,
              200,
              1.5968063872255487],
             'appliances.environment.vacuum': [74580,
              1213,
              1.6264414052024672],
             'kids.toys': [8065, 72, 0.8927464352138872],
             'apparel.shoes': [60211, 331, 0.5497334374117686],
             'appliances.kitchen.refrigerators': [68804,
              866,
              1.2586477530376141],
             'construction.tools.drill': [20947, 20

In [44]:
d2

defaultdict(list,
            {'computers.desktop': [39439, 311, 0.7885595476558735],
             'apparel.shoes.keds': [35708, 0, 0.0],
             'appliances.environment.water_heater': [13233,
              395,
              2.9849618378296685],
             'electronics.smartphone': [1060091, 55155, 5.202855226579605],
             'computers.peripherals.monitor': [10465, 110, 1.051122790253225],
             'computers.notebook': [110126, 1560, 1.4165592139912464],
             'appliances.environment.air_heater': [12525,
              200,
              1.5968063872255487],
             'appliances.environment.vacuum': [74580, 1703, 2.283454009117726],
             'kids.toys': [8065, 40, 0.49597024178549287],
             'apparel.shoes': [60211, 0, 0.0],
             'appliances.kitchen.refrigerators': [68804,
              1237,
              1.7978605895006106],
             'construction.tools.drill': [20947, 154, 0.7351888098534396],
             'auto.accessories.player

In [45]:
dic2 = sorted(d2.items(), key = lambda x: x[1][2], reverse= True)
dic2[:5]

[('electronics.smartphone', [1060091, 55155, 5.202855226579605]),
 ('electronics.audio.headphone', [100622, 5087, 5.055554451312834]),
 ('medicine.tools.tonometer', [1391, 60, 4.313443565780014]),
 ('electronics.video.tv', [104865, 3658, 3.4882944738473274]),
 ('electronics.tablet', [29890, 914, 3.0578788892606226])]

<h2>5.	What is top pair of categories which have high affinity for each other.<h2>

<h4>Based on the above results, Electronics.smartphone and electronics.audio.headphone might be the top pairs of categories which have affinity each other.<h4>